File: src/algorithms/graph/matching/hungarian-bipartite/hungarian-bipartite.test.ts
import { describe, it, expect } from "vitest";

import { hungarianMatching } from "./sources/hungarian-bipartite.ts?fn";

type AdjacencyList = Record<string, string[]>;

describe("hungarianMatching", () => {
  it("finds a perfect matching for a fully-matchable bipartite graph", () => {
    const adjacencyList: AdjacencyList = {
      L1: ["R1", "R2"],
      L2: ["R2", "R3"],
      L3: ["R1", "R3"],
      R1: ["L1", "L3"],
      R2: ["L1", "L2"],
      R3: ["L2", "L3"],
    };
    const result = hungarianMatching(adjacencyList, ["L1", "L2", "L3"], ["R1", "R2", "R3"]);
    expect(Object.keys(result)).toHaveLength(3);
    // Every left node must be matched
    expect(result["L1"]).toBeDefined();
    expect(result["L2"]).toBeDefined();
    expect(result["L3"]).toBeDefined();
    // No two left nodes share the same right node
    const rightValues = Object.values(result);
    expect(new Set(rightValues).size).toBe(rightValues.length);
  });

  it("returns a partial matching when not all left nodes can be matched", () => {
    // Both L1 and L2 only connect to R1 — only one can be matched
    const adjacencyList: AdjacencyList = {
      L1: ["R1"],
      L2: ["R1"],
      R1: ["L1", "L2"],
    };
    const result = hungarianMatching(adjacencyList, ["L1", "L2"], ["R1"]);
    expect(Object.keys(result)).toHaveLength(1);
    const matchedLeft = Object.keys(result)[0]!;
    expect(result[matchedLeft]).toBe("R1");
  });

  it("returns empty matching for a graph with no edges", () => {
    const adjacencyList: AdjacencyList = {
      L1: [],
      L2: [],
      R1: [],
      R2: [],
    };
    const result = hungarianMatching(adjacencyList, ["L1", "L2"], ["R1", "R2"]);
    expect(Object.keys(result)).toHaveLength(0);
  });

  it("matches a single left-right pair correctly", () => {
    const adjacencyList: AdjacencyList = {
      L1: ["R1"],
      R1: ["L1"],
    };
    const result = hungarianMatching(adjacencyList, ["L1"], ["R1"]);
    expect(result["L1"]).toBe("R1");
    expect(Object.keys(result)).toHaveLength(1);
  });

  it("finds augmenting path to re-route an existing match", () => {
    // L1 matches R1 first, then L2 needs R1 and must re-route L1 to R2
    const adjacencyList: AdjacencyList = {
      L1: ["R1", "R2"],
      L2: ["R1"],
      R1: ["L1", "L2"],
      R2: ["L1"],
    };
    const result = hungarianMatching(adjacencyList, ["L1", "L2"], ["R1", "R2"]);
    // Both should be matched (L2 gets R1, L1 re-routes to R2)
    expect(Object.keys(result)).toHaveLength(2);
    const rightValues = Object.values(result);
    expect(new Set(rightValues).size).toBe(2);
  });

  it("handles a one-to-one bipartite graph with guaranteed perfect matching", () => {
    const adjacencyList: AdjacencyList = {
      L1: ["R1"],
      L2: ["R2"],
      L3: ["R3"],
      R1: ["L1"],
      R2: ["L2"],
      R3: ["L3"],
    };
    const result = hungarianMatching(adjacencyList, ["L1", "L2", "L3"], ["R1", "R2", "R3"]);
    expect(result["L1"]).toBe("R1");
    expect(result["L2"]).toBe("R2");
    expect(result["L3"]).toBe("R3");
  });

  it("returns empty matching for an empty graph with no nodes", () => {
    const result = hungarianMatching({}, [], []);
    expect(Object.keys(result)).toHaveLength(0);
  });
});

File: src/algorithms/graph/matching/hungarian-bipartite/__tests__/HungarianBipartite_test.cpp
#include "../sources/HungarianBipartite.cpp"
#include <cassert>
#include <iostream>
#include <set>

int main() {
    // Test 1: perfect matching
    {
        unordered_map<string, vector<string>> adj = {
            {"L1", {"R1", "R2"}}, {"L2", {"R2", "R3"}}, {"L3", {"R1", "R3"}},
            {"R1", {"L1", "L3"}}, {"R2", {"L1", "L2"}}, {"R3", {"L2", "L3"}},
        };
        auto result = HungarianBipartite::hungarianMatching(
            adj, {"L1", "L2", "L3"}, {"R1", "R2", "R3"});
        assert(result.size() == 3);
        set<string> rightValues;
        for (auto& entry : result) rightValues.insert(entry.second);
        assert(rightValues.size() == 3);
    }

    // Test 2: partial matching
    {
        unordered_map<string, vector<string>> adj = {
            {"L1", {"R1"}}, {"L2", {"R1"}}, {"R1", {"L1", "L2"}},
        };
        auto result = HungarianBipartite::hungarianMatching(adj, {"L1", "L2"}, {"R1"});
        assert(result.size() == 1);
        assert(result.begin()->second == "R1");
    }

    // Test 3: no edges → empty matching
    {
        unordered_map<string, vector<string>> adj = {
            {"L1", {}}, {"L2", {}}, {"R1", {}}, {"R2", {}},
        };
        auto result = HungarianBipartite::hungarianMatching(adj, {"L1", "L2"}, {"R1", "R2"});
        assert(result.empty());
    }

    // Test 4: single pair
    {
        unordered_map<string, vector<string>> adj = {{"L1", {"R1"}}, {"R1", {"L1"}}};
        auto result = HungarianBipartite::hungarianMatching(adj, {"L1"}, {"R1"});
        assert(result.at("L1") == "R1");
        assert(result.size() == 1);
    }

    // Test 5: augmenting path
    {
        unordered_map<string, vector<string>> adj = {
            {"L1", {"R1", "R2"}}, {"L2", {"R1"}}, {"R1", {"L1", "L2"}}, {"R2", {"L1"}},
        };
        auto result = HungarianBipartite::hungarianMatching(adj, {"L1", "L2"}, {"R1", "R2"});
        assert(result.size() == 2);
        set<string> rightValues;
        for (auto& entry : result) rightValues.insert(entry.second);
        assert(rightValues.size() == 2);
    }

    // Test 6: one-to-one perfect matching
    {
        unordered_map<string, vector<string>> adj = {
            {"L1", {"R1"}}, {"L2", {"R2"}}, {"L3", {"R3"}},
            {"R1", {"L1"}}, {"R2", {"L2"}}, {"R3", {"L3"}},
        };
        auto result = HungarianBipartite::hungarianMatching(
            adj, {"L1", "L2", "L3"}, {"R1", "R2", "R3"});
        assert(result.at("L1") == "R1");
        assert(result.at("L2") == "R2");
        assert(result.at("L3") == "R3");
    }

    // Test 7: empty graph
    {
        auto result = HungarianBipartite::hungarianMatching({}, {}, {});
        assert(result.empty());
    }

    cout << "All tests passed!" << endl;
    return 0;
}
